### PR TITLE
feat(h705): Introduce a function for conditional logic that evaluates regex

### LINKED ISSUE #705: Introduce a function for conditional logic that evaluates regex
### Goal

Extend the native SurveyJS expression engine with a custom `regexMatch(value, pattern)` function so survey scripters can drive conditional logic (e.g. `visibleIf`, `enableIf`) from regular-expression matches. Regex validation is supported out of the box today; conditional logic based on regex currently requires ad hoc custom functions.

### Scope

**In scope**

- Implement `regexMatch(value, pattern)` — accepts a string value (survey response) and a regex pattern string.
- Register the function via SurveyJS `FunctionFactory` so it is globally available across the survey instance.
- Fail gracefully: return `false` when input is null/undefined or the regex string is syntactically invalid (no console errors or survey crash).
- Validate behavior in both the survey runtime engine and Survey Creator preview.
- Provide end-user documentation.

**Out of scope**

- Custom Survey Creator UI / drag-and-drop widget for this function. Scripters will type `regexMatch(param, param)` into existing expression fields.

### Implementation Notes

Follow the existing `hub/lib/survey-features/*` pattern for registering custom expression functions (e.g. `any-answered/infrastructure/registry.ts`).

Example usage: `regexMatch({question_name}, '^pattern$') == true` in `visibleIf` or `enableIf`.

### Acceptance Criteria

- [ ] **Functional execution.** A survey scripter can use `regexMatch({question_name}, '^pattern$') == true` in `visibleIf` or `enableIf`, and the logic executes correctly on the front end.
- [ ] **Edge case handling.** The application does not throw console errors or crash the survey when `{question_name}` is empty, undefined, or the regex string is malformed.
- [ ] **Test coverage.** The function correctly evaluates:
  - `^\d{4}$` — exactly 4 digits
  - `^01[016789]-?\d{3,4}-?\d{4}$` — Korean mobile phone numbers
  - `^[A-Z]{3}-\d{4}$` — code format (e.g. `ABC-1234`)
  - `^\d{1,3}(\.\d{1,2})?$` — numbers with up to 2 decimal places
  - `^(yes|no)$` — yes/no only (case-insensitive)
- [ ] **Cross-platform support.** Identical behavior in the survey-taking view and Survey Creator preview tab.

### PR DESCRIPTION
**Pull request by orilux**
## Description

- Adds `regexMatch(value, pattern, flags?)` as a custom SurveyJS expression function for regex-based conditional logic (`visibleIf`, `enableIf`, etc.)
- Follows the existing `any-answered` pattern: pure function + `FunctionFactory` registry + React hook, registered globally before survey/Creator initialization
- Hub-only change; no API or Creator UI work

## Behavior

- Returns `true` when the value matches the pattern, `false` otherwise
- Uses `Helpers.isValueEmpty()` so empty strings, `null`, `undefined`, and cleared checkbox `[]` are treated as unanswered
- Coerces scalar non-string values (e.g. numbers) to strings before matching; `0` is a valid answer
- Invalid or missing patterns/flags return `false` without throwing or logging errors
- Optional third parameter supports regex flags (e.g. `'i'` for case-insensitive)

## Wiring

Registers `regexMatch` at all survey runtime and Creator entry points:

- Public survey-taking view
- Submission review runtime
- Form editor (Creator)
- Form preview
- Template preview

## Tests

- Unit tests for match/mismatch, edge cases, AC patterns (4-digit, Korean mobile, `ABC-1234`, decimals, case-insensitive yes/no)
- Integration tests via `survey.runCondition()` and `visibleIf` visibility toggling
- 19 tests passing

## Example

```text
regexMatch({q1}, '^\d{4}$') == true
```

Type single backslashes in the Creator expression box; JSON storage will escape them automatically.

## Related Issues
#705 

## Type of Change
Check all that apply.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Dependency update
- [ ] Refactoring (non-breaking change which improves code quality)
- [ ] Other (please describe)

## Checklist
- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

> [!NOTE]
> - [x] I confirm that my Pull Request follows all the checklist requirements above

## Summary by CodeRabbit

* **New Features**
  * Added regex-based matching support in survey logic, enabling forms and conditions to validate values with regular expressions.
  * Regex matching now works consistently in the editor, preview, public form, and submission experiences.

* **Bug Fixes**
  * Improved handling for empty, invalid, or malformed regex inputs so surveys fail safely instead of throwing errors.
  * Added coverage for more matching scenarios, including case-insensitive patterns and formatted values.

### CODE REVIEW

File: lib/survey-features/regex-match/__tests__/registry.test.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
+import { Model } from 'survey-core';
+import { beforeAll, describe, expect, it } from 'vitest';
+import { registerRegexMatchGlobals } from '../infrastructure/registry';
+
+describe('registerRegexMatchGlobals', () => {
+  beforeAll(() => {
+    registerRegexMatchGlobals();
+    registerRegexMatchGlobals();
+  });
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/regex-match/__tests__/registry.test.ts]: Addressed 🥕

File: lib/survey-features/regex-match/functions.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  const [value, pattern, flags] = params;
+  if (typeof pattern !== 'string' || pattern.length === 0) return false;
+  if (Helpers.isValueEmpty(value)) return false;
+
+  try {
+    const regex = new RegExp(
+      pattern,
+      typeof flags === 'string' ? flags : undefined,
+    );
+    return regex.test(String(value));
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/regex-match/functions.ts]: Addressed 🥕

File: lib/survey-features/regex-match/__tests__/functions.test.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  it('supports optional case-insensitive flags', () => {
+    expect(regexMatchFunction(['YES', '^(yes|no)$', 'i'])).toBe(true);
+    expect(regexMatchFunction(['No', '^(yes|no)$', 'i'])).toBe(true);
+    expect(regexMatchFunction(['maybe', '^(yes|no)$', 'i'])).toBe(false);
+  });
+
+  it('returns false for patterns that risk catastrophic backtracking', () => {
+    // Act & Assert
+    expect(() =>
+      regexMatchFunction(['aaaaaaaaaaaaaaaaaaaa!', '(a+)+$']),
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/regex-match/__tests__/functions.test.ts]: Addressed

File: lib/survey-features/regex-match/__tests__/functions.test.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    ).not.toThrow();
+    expect(regexMatchFunction(['aaaaaaaaaaaaaaaaaaaa!', '(a+)+$'])).toBe(false);
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/regex-match/__tests__/functions.test.ts]: Addressed

File: lib/survey-features/regex-match/__tests__/functions.test.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    expect(regexMatchFunction(['text', '(.*)*$'])).toBe(false);
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/regex-match/__tests__/functions.test.ts]: Addressed

File: customizations/questions/question-registry.ts
Comment (PeterMarinov) [inline customizations/questions/question-registry.ts:1]: Remove this file or this will cause merge conflicts with customers

File: features/forms/ui/preview/preview-form.tsx [hunk 6]
```diff
@@ -30,10 +31,19 @@ const PreviewForm = ({ model, slkVal }: PreviewFormProps) => {
   const [creator, setCreator] = useState<SurveyCreator | null>(null);
   useRichTextEditing(creator);
   useLoopAwareSummaryTableEditing(creator);
+  const { isReady: isExtensionsReady, onCreatorCreated } = useSurveyExtensions({
```
Comment (PeterMarinov) [inline features/forms/ui/preview/preview-form.tsx:34]: Oops, seems we have forgotten to add the extensions here. Good that you have addressed this

File: features/submissions/ui/shared/use-survey-model.hook.ts [hunk 7]
```diff
@@ -29,13 +30,23 @@ export function useSurveyModel(
     bindToSurvey: bindQuestionLoops,
   } = useQuestionLoops();
   const { initGlobals: initAnyAnsweredGlobals } = useAnyAnswered();
+  const { isReady: isExtensionsReady } = useSurveyExtensions({
```
Comment (PeterMarinov) [inline features/submissions/ui/shared/use-survey-model.hook.ts]: Review the use of useSurveyExtensions. I believe, it should not be here

`ViewSubmissionCore` and `EditSubmissionCore` are both adding the extensions via `useSurveyExtensions`. Then they pass them to the `SubmissionSurvey`, which uses the  `useSurveyModel` so you seem to be loading the extensions twice.

Is there a reason to add the `useSurveyExtensions` here? Please review and I recommend cleaning it up

File: lib/survey-features/regex-match/functions.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import { Helpers } from 'survey-core';
+import {
+  isSafeRegexPattern,
+  MAX_REGEX_VALUE_LENGTH,
+} from './pattern-safety';
+
+export function regexMatchFunction(params: unknown[]): boolean {
```
Comment (PeterMarinov) [inline lib/survey-features/regex-match/functions.ts:18]: `regexMatch()` runs synchronous RegExp.test() in the browser via SurveyJS FunctionFactory. Business users can supply arbitrary patterns in visibleIf / enableIf leading to blocking main thread.

Current mitigations in pattern-safety.ts + functions.ts are a good start (length caps, nested-quantifier heuristic, invalid-syntax guard) but are best-effort, not a full ReDoS guarantee.

I recommend that we merge  current implementation as MVP with documented limits and a plan we can address as follow-up:

#### Recommended hardening plan
please review and make a plan if you agree with this. Then mark the comment as resolved

##### Phase 1 — Strengthen static checks (low effort, high value)

- Add a dedicated analyzer (safe-regex or recheck) inside isSafeRegexPattern().
- Keep existing length limits and nested-quantifier rule as defense-in-depth.
- Add tests for known ReDoS fixtures beyond (a+)+$ / (.*)*$.
- Document “safe pattern” guidance for form authors (avoid .*, unbounded +/*, deep nesting).

##### Phase 2 — Designer-time validation (prevent bad patterns at source)

- On form save / JSON validation, scan expressions for regexMatch(...) and reject or warn on unsafe patterns.
- Surface errors in Creator (logic tab / JSON editor) before publish.
- Optional: telemetry counter for runtime rejections (isSafeRegexPattern returned false) to find false positives, so we can track cases with bad regEx - e.g. PostHog event tracking or Telemetry logger

**Review by PeterMarinov — CHANGES_REQUESTED** [review]
Requested changes to avoid double registration of the extensions using the `use-survey-model.hook`

File: customizations/questions/question-registry.ts
Comment (orilux) [reply customizations/questions/question-registry.ts:1]: Removed

File: features/forms/ui/preview/preview-form.tsx (hunk 6 above)
Comment (orilux) [reply features/forms/ui/preview/preview-form.tsx:34]: It happens to the best of us.

File: features/submissions/ui/shared/use-survey-model.hook.ts (hunk 7 above)
Comment (orilux) [reply features/submissions/ui/shared/use-survey-model.hook.ts]: No reason. This was a mistake. Corrected.

File: lib/survey-features/regex-match/functions.ts (hunk 8 above)
Comment (orilux) [reply lib/survey-features/regex-match/functions.ts:18]: Issue and plan created. https://github.com/endatix/endatix-hub/issues/712

**Review by PeterMarinov — APPROVED** [review]
Approved
